dopamine/utils/threading_utils.py

```
@@ -0,0 +1,100 @@
+# coding=utf-8
+# Copyright 2018 The Dopamine Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Decorator to decouple object attributes and make them local to threads."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import threading
+
+
+def _get_internal_name(name):
+  """Returns the thread local name of an attribute."""
+  return '__' + name + '_' + str(threading.current_thread().ident)
+
+
+def _get_default_value_name(name):
+  """Returns the global default value of an attribute."""
+  return name + '_default'
+
+
+def _add_property(cls, attr_name):
+  """Adds a property to a given class.
+
+  The a setter, getter and deleter are added to the given class with to the
+  provided attribute name.
+  These methods actually apply to an intenal variable that is thread-specific.
+  Hence the result of these methods depends on the local thread.
+
+  Note that when the getter is called and the local attribute is not found the
+  getter will initialize the local value to the global default value.
+  See `initialize_local_attributes` for more details.
+
+  Args:
+    cls: A class to add the poperty to.
+    attr_name: str, name of the property to create.
+  """
+  def _set(self, val):
+    setattr(self, _get_internal_name(attr_name), val)
+
+  def _get(self):
+    if not hasattr(self, _get_internal_name(attr_name)):
+      _set(self, getattr(self, _get_default_value_name(attr_name)))
```

Are you comfortable with the exception that'll be raised by `getattr` if `_get` is called before any `_set` and without any `initialize_local_arguments`?

---

I am comfortable with the exception being raised - that's the behavior that I think is adapted.
I though think we need to raise a specific error with more information on what is actually the problem and how to use the API.

Updated with a custom error.

---

Good call.